// See README.txt for information and build instructions.
//
// Note: START and END tags are used in comments to define sections used in
// tutorials.  They are not part of the syntax for Protocol Buffers.
//
// To get an in-depth walkthrough of this file and the related examples, see:
// https://developers.google.com/protocol-buffers/docs/tutorials

// [START declaration]
syntax = "proto3";
package poses;

// [END declaration]

// [START csharp_declaration]
option csharp_namespace = "Polaris.Protobuf";
// [END csharp_declaration]

// [START messages]

message Vector3DTO {
  float x = 1;
  float y = 2;
  float z = 3;
}

message KeyframeDTO {
  float time = 1;
  float value = 2;
  float inTangent = 3;
  float outTangent = 4;
}

message KeypointEstimation {
  float x = 101;
  float y = 102;
  float z = 103;
  float p = 201;
}

message KeypointEstimationFrames {
  string id = 101;
  repeated KeypointEstimation frames = 201;
}

message PoseEstimationFrames {
  repeated KeypointEstimationFrames keypoints = 201;
}

message PoseEstimationResult {
  PoseEstimationFrames poseEstimations = 101;
}

message MoveDTO
{
  repeated TransformCurvesDTO transformCurves = 1;
  float length = 2;
}

message TransformCurvesDTO 
{
  string transformPath = 1;
  Vector3CurvesDTO positionCurves = 2;
  QuaternionCurvesDTO rotationCurves = 3;
  
}

message Vector3CurvesDTO 
{	
  repeated KeyframeDTO x = 1;
  repeated KeyframeDTO y = 2;
  repeated KeyframeDTO z = 3;
}

message QuaternionCurvesDTO 
{	
  repeated KeyframeDTO x = 1;
  repeated KeyframeDTO y = 2;
  repeated KeyframeDTO z = 3;
  repeated KeyframeDTO w = 4;
}

message TransformPoseDTO
{
  string transformPath = 1;
  Vector3DTO pos = 2;
  Vector3DTO rot = 3;
}

message PoseDTO
{
  repeated TransformPoseDTO transforms = 1;
}

message AnimatablePoseDTO
{
  PoseDTO bones = 1;
  PoseDTO controls = 2;
}

message EditorMovePoseDTO
{
  int32 frame = 1;
  PoseDTO pose = 2;
}

message EditorMoveMetaDTO
{
  int32 framesPerSecond = 1;
  int32 framesTotal = 2;
}

message EditorMoveDTO
{
  EditorMoveMetaDTO metadata = 1;
  repeated EditorMovePoseDTO poses = 2;
}
// [END messages]
